ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ObjectDiffusion/Smoke.hs

```diff
@@ -0,0 +1,315 @@
+        atomically $ modifyTVar poolContent (++ objects)
+        return ()
+    }
+
+mkMockPoolInterfaces ::
+  MonadSTM m =>
+  [SmokeObject] ->
+  m
+    ( ObjectPoolReader SmokeObjectId SmokeObject Int m
+    , ObjectPoolReader SmokeObjectId SmokeObject Int m
+    , ObjectPoolWriter SmokeObjectId SmokeObject m
+    )
+mkMockPoolInterfaces objects = do
+  outboundPool <- newObjectPool objects
+  inboundPool <- newObjectPool []
+
+  let outboundPoolReader = makeObjectPoolReader outboundPool
+      inboundPoolReader = makeObjectPoolReader inboundPool
+      inboundPoolWriter = makeObjectPoolWriter inboundPool
+
+  return (outboundPoolReader, inboundPoolReader, inboundPoolWriter)
+
+{-------------------------------------------------------------------------------
+  Main properties
+-------------------------------------------------------------------------------}
+
+-- Protocol constants
+
+numObjectIdsToAck :: NumObjectIdsToAck
+numObjectIdsToAck = NumObjectIdsToAck 10
+
+nodeToNodeVersion :: NodeToNodeVersion
+nodeToNodeVersion = NodeToNodeV_14
+
+prop_smoke_init_inbound :: ListWithUniqueIds SmokeObject idTy -> Property
+prop_smoke_init_inbound (ListWithUniqueIds objects) =
+  prop_smoke_object_diffusion objects mkOutboundAsync mkInboundAsync (mkMockPoolInterfaces objects)
+ where
+  mkOutboundAsync outbound outboundChannel tracer =
+    async $
+      ( ()
+          <$ runPeer
+            ((\x -> "Outbound (Server): " ++ show x) `contramap` tracer)
+            codecObjectDiffusionId
+            outboundChannel
+            (objectDiffusionOutboundServerPeer outbound)
+      )
+
+  mkInboundAsync inbound inboundChannel tracer =
+    async $
+      ( ()
+          <$ runPipelinedPeer
+            ((\x -> "Inbound (Client): " ++ show x) `contramap` tracer)
+            codecObjectDiffusionId
+            inboundChannel
+            (objectDiffusionInboundClientPeerPipelined inbound)
+      )
+
+prop_smoke_init_outbound :: ListWithUniqueIds SmokeObject SmokeObjectId -> Property
+prop_smoke_init_outbound (ListWithUniqueIds objects) =
+  prop_smoke_object_diffusion objects mkOutboundAsync mkInboundAsync (mkMockPoolInterfaces objects)
+ where
+  mkOutboundAsync outbound outboundChannel tracer =
+    async $
+      ( ()
+          <$ runPeer
+            ((\x -> "Outbound (Client): " ++ show x) `contramap` tracer)
+            codecObjectDiffusionId
+            outboundChannel
+            (objectDiffusionOutboundClientPeer outbound)
+      )
+
+  mkInboundAsync inbound inboundChannel tracer =
+    async $
+      ( ()
+          <$ runPipelinedPeer
+            ((\x -> "Inbound (Server): " ++ show x) `contramap` tracer)
+            codecObjectDiffusionId
+            inboundChannel
+            (objectDiffusionInboundServerPeerPipelined inbound)
+      )
+
+--- The core logic of the smoke test is shared between the generic smoke tests for ObjectDiffusion, and the ones specialised to PerasCert/PerasVote diffusion
+prop_smoke_object_diffusion ::
```

Comment:
This can probably be polymorphic in the monad (with an prop_smoke_object_diffusion constraint)?